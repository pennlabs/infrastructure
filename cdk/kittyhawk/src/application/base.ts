import { Construct } from "constructs";
import { removeSubdomain } from "..";
import { Certificate } from "../certificate";
import { Deployment, DeploymentProps } from "../deployment";
import { Ingress, IngressProps } from "../ingress";
import { Service } from "../service";
import { ServiceAccount } from "../serviceaccount";

export interface ApplicationProps {
  /**
   * Ingress configuration
   */
  readonly ingress?: IngressProps;

  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || "undefined_release";
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name,
      });
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      port: props.port,
      ...(props.createServiceAccount
        ? { serviceAccountName: release_name }
        : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      const alreadyCreatedCertificates = new Set<string>();

      props.ingress.rules.forEach((rules) => {
        const finalDomain: string = removeSubdomain(
          rules.host,
          rules.isSubdomain ?? false
        );
        if (!alreadyCreatedCertificates.has(finalDomain)) {
          new Certificate(this, fullname, rules);
          alreadyCreatedCertificates.add(finalDomain);
        }
      });
    }
  }
}
