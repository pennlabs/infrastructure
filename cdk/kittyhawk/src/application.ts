import { Construct } from 'constructs';
import { Deployment, DeploymentProps } from './deployment';
import { Ingress, IngressProps } from './ingress';
import { Service, ServiceProps } from './service'
import { Certificate } from './certificate';

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have 
 * property names that conflict with each other. Typescript may not throw an error and it 
 * could cause problems.
 */
export interface ApplicationProps extends IngressProps, DeploymentProps,
  ServiceProps { }


export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`

    new Service(this, fullname, props);

    new Deployment(this, fullname, props);

    if (props.ingress) {
      new Ingress(this, fullname, props)

      new Certificate(this, fullname, props)
    }
  }
}

/**
 * Helper function that first checks to make sure that the environment variable array 
 * doesn't already contain the env var, then inserts it into the array.
 * @param envArray array of environment variables to insert into 
 * @param envKey name of the environment variable to insert
 * @param envValue value of the environment variable
 */
function insertIfNotPresent(envArray: { name: string, value: string }[], envKey: string, envValue: any) {
  const envSettingsModule = envArray?.filter(env => (env.name === envKey));
  if (envSettingsModule?.length > 0) {
    throw new Error(`${envKey} should not be redefined as an enviroment variable.`)
  }
  envArray.push({ name: envKey, value: envValue });

}
export interface DjangoApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;
}

export class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Have to be careful here with references when mutating things
    let djangoExtraEnv = Array.from(props.extraEnv || []);

    // Insert DJANGO_SETTINGS_MODULE and DOMAIN
    insertIfNotPresent(djangoExtraEnv, 'DJANGO_SETTINGS_MODULE', props.djangoSettingsModule)
    insertIfNotPresent(djangoExtraEnv, 'DOMAIN', props.domain)

    // Configure the ingress using ingressPaths.
    const djangoIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: djangoExtraEnv,
      ingress: djangoIngress,
    });
  }
}

export interface ReactApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {

    // Have to be careful here with references when mutating things
    let reactExtraEnv = Array.from(props.extraEnv || []);

    // Insert DOMAIN and PORT as env vars.
    insertIfNotPresent(reactExtraEnv, 'DOMAIN', props.domain)
    insertIfNotPresent(reactExtraEnv, 'PORT', props.portEnv || '80')

    // Configure the ingress using ingressPaths.
    const reactIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: reactExtraEnv,
      ingress: reactIngress,
    });
  }
}

export class RedisApplication extends Application {
  constructor(scope: Construct, appname: string, redisProps: Partial<ApplicationProps>) {
    super(scope, appname, {
      ...redisProps,
      image: 'redis',
      tag: redisProps.tag || '6.0',
      port: redisProps.port || 6379,
    });
  }
}
