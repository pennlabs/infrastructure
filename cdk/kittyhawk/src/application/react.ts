import { Construct } from "constructs";
import { DeploymentProps } from "../deployment";
import { HostRules, IngressProps } from "../ingress";
import { Application } from "./base";

export interface ReactApplicationProps {
  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Host is the domain the application runs on,
   * isSubdomain is true if the domain should be treated as a subdomain for certificate purposes.
   * paths is the list of paths to expose the application on.
   * See the certificate documentation for more details.
   *
   */
  readonly domain: HostRules;

  /**
   * Optional ingressProps to override the default ingress props.
   */
  readonly ingressProps?: IngressProps;

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {
    // Now, we ensure there are no duplicate env variables, even if they redefine it
    const reactExtraEnv = [
      ...new Set([
        ...(props.deployment?.env || []),
        { name: "DOMAIN", value: props.domain.host },
        { name: "PORT", value: props.portEnv || "80" },
      ]),
    ];

    // If everything passes, construct the Application.
    super(scope, appname, {
      port: props.port ?? 80,
      deployment: {
        ...props.deployment,
        env: reactExtraEnv,
      },
      ingress: {
        rules: [props.domain],
        ...props.ingressProps,
      },
      createServiceAccount: props.createServiceAccount,
    });
  }
}
