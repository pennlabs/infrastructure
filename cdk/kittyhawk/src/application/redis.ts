import { Construct } from "constructs";
import { DeploymentProps } from "../deployment";
import { Application } from "./base";
import {
  KubePersistentVolume,
  KubePersistentVolumeClaim,
  Quantity,
} from "../imports/k8s";

export interface RedisApplicationProps {
  /**
   * Deployment configuration
   * @default {image: 'redis', tag: '6.0', port: 6379}
   */
  readonly deployment?: Partial<DeploymentProps>;

  /**
   * Port to expose the application on.
   * @default 6379
   */
  readonly port?: number;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;

  /**
   * Persists data to a volume, using a ConfigMap to decide where to mount it.
   */
  persistData?: boolean;
}

export class RedisApplication extends Application {
  constructor(
    scope: Construct,
    appname: string,
    redisProps: RedisApplicationProps
  ) {
    const CONFIG_MAP_NAME = "redis-config";

    if (redisProps.persistData) {
      new KubePersistentVolume(scope, `${appname}-pv`, {
        metadata: {
          name: `${appname}-pv`,
        },
        spec: {
          storageClassName: `${appname}-redis-storage`,
          accessModes: ["ReadWriteMany"], // TODO: WHAT IS THIS
          capacity: {
            storage: Quantity.fromString("1Gi"),
          },
          hostPath: {
            path: `/${appname}/redis`,
          },
        },
      });
      new KubePersistentVolumeClaim(scope, `${appname}-pvc`, {
        metadata: {
          name: `${appname}-pvc`,
        },
        spec: {
          storageClassName: `${appname}-redis-storage`,
          accessModes: ["ReadWriteMany"], // TODO: WHAT IS THIS
          resources: {
            requests: {
              storage: Quantity.fromString("1Gi"),
            },
          },
        },
      });
    }

    super(scope, appname, {
      deployment: {
        image: redisProps.deployment?.image ?? "redis",
        tag: redisProps.deployment?.tag ?? "6.0",
        volumeMounts: [
          ...(redisProps.persistData
            ? [
                {
                  name: "data",
                  mountPath: "/redis-master-data",
                },
                {
                  name: "config",
                  mountPath: "/redis-master",
                },
              ]
            : []),
          ...(redisProps.deployment?.volumeMounts ?? []),
        ],
        volumes: [
          ...(redisProps.persistData
            ? [
                {
                  name: "data",
                  persistentVolumeClaim: {
                    claimName: `${appname}-pvc`,
                  },
                },
                {
                  name: "config",
                  configMap: {
                    name: CONFIG_MAP_NAME,
                    items: [
                      {
                        key: "redis-config",
                        path: "redis.conf",
                      },
                    ],
                  },
                },
              ]
            : []),
          ...(redisProps.deployment?.volumes ?? []),
        ],
        ...redisProps.deployment,
      },
      port: redisProps.port ?? 6379,
      createServiceAccount: redisProps.createServiceAccount,
    });
  }
}
