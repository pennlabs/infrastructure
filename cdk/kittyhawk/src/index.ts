
export * from './application';
export * from './autoscaler';
export * from './certificate';
export * from './chart';
export * from './container';
export * from './cronjob';
export * from './deployment';
export * from './ingress';
export * from './service';
