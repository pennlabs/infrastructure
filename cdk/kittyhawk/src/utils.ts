export type NonEmptyArray<T> = [T, ...T[]];

export function nonEmptyMap<T, U>(
  arr: NonEmptyArray<T>,
  callbackfn: (value: T) => U
): NonEmptyArray<U> {
  return arr.map(callbackfn) as any as NonEmptyArray<U>;
}
