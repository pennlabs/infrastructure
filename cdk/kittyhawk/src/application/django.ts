import { Construct } from 'constructs';
import { Application } from '.';
import { DeploymentProps } from '../deployment';
import { HostRules, IngressProps } from '../ingress';
import { NonEmptyArray, nonEmptyMap } from '../utils';

export interface DjangoApplicationProps {
  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Array of domain(s). Host is the domain the application runs on,
   * and isSubdomain is true if the domain should be treated as a subdomain for certificate purposes.
   * See the certificate documentation for more details.
   */
  readonly domains: NonEmptyArray<{ host: string; isSubdomain?: boolean }>;

  /**
   * Just the list of paths passed to the ingress since we already know the host. Optional.
   *
   * @default undefined
   */
  readonly ingressPaths?: string[];

  /**
   * Optional ingressProps to override the default ingress props.
   */
  readonly ingressProps?: IngressProps;

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}


export default class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Now, we ensure there are no duplicate env variables, even if they redefine it
    const djangoExtraEnv = [...new Set([
      ...props.deployment?.env || [],
      { name: 'DJANGO_SETTINGS_MODULE', value: props.djangoSettingsModule },
      { name: 'DOMAIN', value: nonEmptyMap(props.domains, (h => h.host)).join() },
    ])];

    // Configure the ingress using ingressPaths if ingressPaths is defined.
    const djangoIngress: HostRules[] = props.domains?.map(h => ({
      host: h.host,
      paths: props.ingressPaths || [],
      isSubdomain: h.isSubdomain ?? false,
    }));

    // If everything passes, construct the Application.
    super(scope, appname, {
      port: props.port ?? 80,
      deployment: {
        ...props.deployment,
        env: djangoExtraEnv,
      },
      ingress: { 
        rules: djangoIngress as NonEmptyArray<HostRules>,
        ...props.ingressProps,
       },
      createServiceAccount: props.createServiceAccount,
    });
  }
}