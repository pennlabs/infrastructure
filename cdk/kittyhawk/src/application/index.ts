import { Construct } from 'constructs';
import { Certificate } from '../certificate';
import { Deployment, DeploymentProps } from '../deployment';
import { Ingress, IngressProps } from '../ingress';
import { Service } from '../service';
import { ServiceAccount } from '../serviceaccount';
import { nonEmptyMap } from '../utils';

import DjangoApplication from './django';
import ReactApplication from './react';
import RedisApplication from './redis';

export { DjangoApplication };
export { ReactApplication };
export { RedisApplication };

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have
 * property names that conflict with each other. Typescript may not throw an error and it
 * could cause problems.
 */
export interface ApplicationProps {
  readonly ingress?: IngressProps;
  readonly deployment: DeploymentProps;
  readonly port?: number;
  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name
      })
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      ...(props.createServiceAccount ? { serviceAccountName: release_name } : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      nonEmptyMap(props.ingress.rules, rule => {
        return new Certificate(this, fullname, rule);
      });
    }
  }
}