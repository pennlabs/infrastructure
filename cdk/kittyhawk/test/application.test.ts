import { Construct } from 'constructs';
import { Application, DjangoApplication, ReactApplication } from '../src/application'
import { chartTest, failingTest} from './utils'


export function buildTagOverrideChart(scope: Construct) {

  /** Overrides the image tag set as env var **/
  new Application(scope, 'serve', {
    image: 'pennlabs/website',
    tag: 'latest',
  })
}


export function buildFailingDjangoChart(scope: Construct) {

  /** Django Duplicated DOMAIN Env should fail **/
  new DjangoApplication(scope, 'platform', {
    image: 'pennlabs/platform',
    domain: 'platform.pennlabs.org',
    djangoSettingsModule: 'Platform.settings.production',
    extraEnv: [ { name: 'DOMAIN', value: 'platform.pennlabs.org' }],
    ingressPaths: ['/'],
  })
}
  
export function buildFailingReactChart(scope: Construct) {
  
  /** React Duplicated DOMAIN Env should fail **/
  new ReactApplication(scope, 'react', {
    image: 'pennlabs/penn-clubs-frontend',
    replicas: 2,
    domain: 'pennclubs.com',
    ingressPaths: ['/'],
    extraEnv: [ { name: 'DOMAIN', value: 'pennclubs.com' }],
    portEnv: '80',
  })
}

test('Tag Override', () => chartTest(buildTagOverrideChart));
  
test('Django Application -- Failing', () => failingTest(buildFailingDjangoChart));
  
test('React Application -- Failing', () => failingTest(buildFailingReactChart));