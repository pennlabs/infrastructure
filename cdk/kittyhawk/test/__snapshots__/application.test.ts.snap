// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Override 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-serve",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-serve",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-serve",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-serve",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-serve",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-serve",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "pennlabs/website:latest",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
]
`;
